File: Teaching/Fractals.F15/euler.py
# Program that uses Euler's method to numerically solve an ODE
#
# David Feldman
# January 13, 2013

from math import exp
from numpy import arange, empty
from pylab import plot,xlabel,ylabel,show,legend

# This will solve ODE's of the form y' = F(y,t)

# **************************************
def F(y,t):
    return -0.2*(y-20)

y_0 = 80
a = 0.0 # starting t value
b = 50.0 # ending t value
n = 100 # number of steps
# **************************************

delta_t = (b-a)/n # stepsize for Euler

tpoints = arange(a,b,delta_t) #tpoints is an array
ypoints = [] #ypoints is a list
exact_ypoints = [] 

y = y_0
for t in tpoints:
    ypoints.append(y)
    y = y + delta_t*F(y,t)

# for this ODE we can get an exact solution.
# let's plot it along with the Euler solution
# make the tpoints for the exact solution
dt = (b-a)/1000 
tpoints2 = arange(a, b, dt)
for t in tpoints2:
    exact_ypoints.append(20 + 60*exp(-0.2*t))    


plot(tpoints,ypoints,label="Euler")
plot(tpoints2,exact_ypoints,label="Exact") #plot exact solution
xlabel("t")
ylabel("y(t)")
legend()
show()

File: KIST/Computational_Physics/euler.m
function [ ] = euler(f,trange,y0,n)
% this code is based on a program written by Todd Young
% and downloaded from http://www.math.ohiou.edu/courses/matlab/
% modified by David Feldman.  28 December 2011
% for use in the Computational Physics class at KIST.
% This function determine approximate solutions to the 
% differential equation y' = f(t,y) with initial condition y(a) = y0
% on the interval [a,b] using n steps of Euler's method.

% **************** Initialze some Variables ***************
a = trange(1); % a is the starting t value
b = trange(2); % b is the ending t value
Delta_T = (b-a)/n;        % step size
t = a; y = y0;      % t and y are the current variables
T = a; Y = y0';      % T and Y will record all steps
% ***************** Begin Main Euler Loop *****************
for i = 1:n
    y = y + Delta_T*f(t,y); % the main Euler step
    t = a + i*Delta_T; % increase the value of t by Delta_t
    T = [T; t];
    Y = [Y; y'];
end
% plot the solution
plot(T,Y)
title('solution using Euler method')
xlabel('t')
ylabel('y(t)')
% display the numerical values:
% comment the below lines if you do not want to see 
% the numerical values that matlab computed
printf("\nt \t\ty(t)\n")
for i = 1:n+1
    printf("%f\t%f\n",T(i),Y(i))
end

% USAGE
% To use this you will need to specify the function f(y,t).
% Do so via the inline function command.  Let f be the name
% of the function.  Then, type the command:
%               myeuler(f, trange,y0,n)
% trange is the range of t values.  If you wanted the solve the equation
% starting at 3 and ending at 5, trange = [3,5].  
% y0 is the initial value.  It is the value of y(t) when t is the first
% value in trange.  
% n is the number of steps.  matlab then computes the step size delta t.

% OTHER TECHNICAL NOTES
% y must be a column vector and f must produce column vectors
% The dimension of y0 must match that of y and f
% T will be a n+1 column vector contianing the times
% Y will be a (n+1) by d matrix where d is the length of y